authorizations/authorization-ranger/src/main/java/org/apache/gravitino/authorization/ranger/RangerAuthorizationPlugin.java

```diff
@@ -0,0 +1,1015 @@
+
+      List<String> nsMetadataObject =
+          Lists.newArrayList(SecurableObjects.DOT_SPLITTER.splitToList(metadataObject.fullName()));
+      if (nsMetadataObject.size() > 4) {
+        // The max level of the securable object is `catalog.db.table.column`
+        throw new RuntimeException("The securable object than 4");
+      }
+      nsMetadataObject.remove(0); // remove `catalog`
+
+      for (int i = 0; i < nsMetadataObject.size(); i++) {
+        RangerPolicy.RangerPolicyResource policyResource =
+            new RangerPolicy.RangerPolicyResource(nsMetadataObject.get(i));
+        policy
+            .getResources()
+            .put(
+                i == 0
+                    ? RangerDefines.RESOURCE_DATABASE
+                    : i == 1 ? RangerDefines.RESOURCE_TABLE : RangerDefines.RESOURCE_COLUMN,
+                policyResource);
+      }
+
+      RangerPolicy finalPolicy = policy;
+      ownerPrivileges.stream()
+          .forEach(
+              ownerPrivilege -> {
+                RangerPolicy.RangerPolicyItem policyItem = new RangerPolicy.RangerPolicyItem();
+                policyItem
+                    .getAccesses()
+                    .add(new RangerPolicy.RangerPolicyItemAccess(ownerPrivilege));
+                if (newOwner != null) {
+                  if (newOwner.type() == Owner.Type.USER) {
+                    policyItem.getUsers().add(newOwner.name());
+                  } else {
+                    policyItem.getGroups().add(newOwner.name());
+                  }
+                }
+                finalPolicy.getPolicyItems().add(policyItem);
+              });
+    }
+    try {
+      if (policy.getId() == null) {
+        rangerClient.createPolicy(policy);
+      } else {
+        rangerClient.updatePolicy(policy.getId(), policy);
+      }
+    } catch (RangerServiceException e) {
+      throw new RuntimeException(e);
+    }
+
+    return Boolean.TRUE;
+  }
+
+  /**
+   * Because one Ranger policy maybe contain multiple Gravitino securable objects, <br>
+   * So we need to find the corresponding policy item mapping to set the user.
+   */
+  @Override
+  public Boolean onGrantedRolesToUser(List<Role> roles, User user) throws RuntimeException {
+    // If the user does not exist, then create it.
+    onUserAdded(user);
+
+    AtomicReference<Boolean> execResult = new AtomicReference<>(Boolean.TRUE);
+    roles.stream()
+        .forEach(
+            role -> {
+              createRangerRoleIfNotExists(role.name());
+              GrantRevokeRoleRequest grantRevokeRoleRequest =
+                  createGrantRevokeRoleRequest(role.name(), user.name(), null);
+              try {
+                rangerClient.grantRole(rangerServiceName, grantRevokeRoleRequest);
+              } catch (RangerServiceException e) {
+                // ignore exception, support idempotent operation
+                LOG.warn("Grant role to user failed!", e);
+              }
+
+              role.securableObjects().stream()
```

Comment:
I think that the `grantRolesToUser` only needs to call `rangerClient.grantRole` to modify the relationship between role and user. It has nothing to do with `policy`, we shouldn't modify the `policy`.

Reply:
I know what you mean.
The First step is to add a user to this role. But we also need to check and add this role to the policy.
otherwise, the role may not be associated with policy item.

Reply:
You are right.
I modified these code.